### PR TITLE
Form get errors

### PR DESCRIPTION
**Pull request by VincentLanglet**

### CODE REVIEW

**Comment by VincentLanglet** [comment]
Hi @seferov, this is the equivalent of the phpstan dynamic return type https://github.com/phpstan/phpstan-symfony/pull/228

I started from
https://github.com/psalm/psalm-plugin-symfony/pull/242
to have the CI green.

File: tests/acceptance/acceptance/forms/FormGetErrors.feature [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+          foreach ($form->getErrors(true, false) as $error5) {
+              /** @psalm-trace $error5 */
+          }
+
+          foreach ($form->getErrors(false, true) as $error6) {
+              /** @psalm-trace $error6 */
+          }
+
+          foreach ($form->getErrors(true, false) as $error7) {
```
Comment (seferov) [inline tests/acceptance/acceptance/forms/FormGetErrors.feature:39]: looks good to me but I don't understand how error5 (L:31) is different than this one?

Comment (VincentLanglet) [reply tests/acceptance/acceptance/forms/FormGetErrors.feature:58]: Nice catch ; seems like this is not working and the test is wrong.

Every error are still Symfony\Component\Form\FormError\|Symfony\Component\Form\FormErrorIterator
and the test is only checking it contains `Symfony\Component\Form\FormError`.

I'll update the test

**Comment by VincentLanglet** [comment]
Any idea was could be wrong with my code @orklah ?

When I add dump inside the return type provider, it's executed and returning
```
return new Type\Union([
            new Type\Atomic\TGenericObject(FormErrorIterator::class, [
                new Type\Union([new TNamedObject(FormError::class)]),
            ]),
        ]);
```

**Comment by orklah** [comment]
Can you give me more details please? What are you trying to do, what is the issue you encounter? Note: I don't use symfony a lot so don't assume I know what those classes do :)

**Comment by VincentLanglet** [comment]
> Can you give me more details please? What are you trying to do, what is the issue you encounter? Note: I don't use symfony a lot so don't assume I know what those classes do :)

Sure, 

The method `FormInterface::getErrors` is returning a FormErrorIterator which is an Iterator of FormError|FormErrorIterator.

When you call `getErrors(true, false)` you really get a `FormErrorIterator<FormError|FormErrorIterator>`.
But in every other case, getErrors is returning a `FormErrorIterator<FormError>.`, so you should be able to iterator on the FormErrorIterator and call `getMessage` (which is a method of a FormError) on every item.

So I tried to introduce a dynamic return type provider, which seems to return the correct value according to my dumps. I even added a stub for FormErrorIterator to be sure it's a generic for the items it iterate over. And still, 
```
 foreach ($form->getErrors() as $error1) {
              $messages[] = $error1->getMessage();
          }
```
is reported as an error.

**Comment by orklah** [comment]
I can't see your Trace in the CI, but the fact Psalm emits a `MixedAssignment` is not a good sign. It may be because of the recursion on `@template T of FormError|FormErrorIterator`

Can you tell me what the Trace is returning on your side?

**Comment by VincentLanglet** [comment]
With 
```
 foreach ($form->getErrors() as $error1) {
              /** @psalm-trace $error1 */
              $messages[] = $error1->getMessage();
          }
```
I had `$error1: FormError|FormErrorIterator`

I didn't try to trace `$foo = $form->getErrors()`, but I will.

**Comment by orklah** [comment]
yeah, please do. $error1 should definitely not be `FormError|FormErrorIterator`

**Comment by VincentLanglet** [comment]
```
$errors = $form->getErrors();
/** @psalm-trace $errors */

foreach ($errors as $error6) {
    /** @psalm-trace $error6 */
    $messages[] = $error6->getMessage();
}
```
gives
```
$errors: Symfony\Component\Form\FormErrorIterator<Symfony\Component\Form\FormError>
$error6: Symfony\Component\Form\FormError|Symfony\Component\Form\FormErrorIterator<Symfony\Component\Form\FormError|Symfony\Component\Form\FormErrorIterator<Symfony\Component\Form\FormError|Symfony\Component\Form\FormErrorIterator>> 
```

I discovered I think it was related to the definition of the current method
```
    /**
     * Returns the current element of the iterator.
     *
     * @return FormError|self An error or an iterator containing nested errors
     */
    public function current()
    {
        return current($this->errors);
    }
```
So I overrade the phpdoc in the stub to
```
    /**
     * @return T
     */
    public function current();
```

**Comment by VincentLanglet** [comment]
I'd like the confirmation by @orklah I did the right fix but I think this is ready to be merged @seferov :)

**Comment by orklah** [comment]
That's what I would have done too at least :)

**Review by seferov — APPROVED** [review]
thank you

**Pull request merged**